File: src/services/geocoding.ts
import type { GeoLocation } from '../types';
import { cacheSet, cacheGet, TTL, CACHE_KEYS } from './cache';

const BASE_URL = 'https://geocoding-api.open-meteo.com/v1';

export async function searchLocations(query: string): Promise<GeoLocation[]> {
  if (!query.trim()) return [];

  const cacheKey = query.toLowerCase().trim();
  const cached = cacheGet<GeoLocation[]>(CACHE_KEYS.GEOCODING, cacheKey, TTL.GEOCODING);
  if (cached) return cached;

  const params = new URLSearchParams({
    name: query,
    count: '10',
    language: 'fr',
    format: 'json',
  });

  const response = await fetch(`${BASE_URL}/search?${params}`);
  if (!response.ok) throw new Error(`Erreur géocodage: ${response.status}`);

  const json = await response.json();
  const results: GeoLocation[] = json.results ?? [];
  cacheSet(CACHE_KEYS.GEOCODING, cacheKey, results);
  return results;
}

export async function reverseGeocode(lat: number, lng: number): Promise<GeoLocation | null> {
  const cacheKey = `${lat.toFixed(3)}_${lng.toFixed(3)}`;
  const cached = cacheGet<GeoLocation>(CACHE_KEYS.GEOCODING, cacheKey, TTL.GEOCODING);
  if (cached) return cached;

  const params = new URLSearchParams({
    latitude: String(lat),
    longitude: String(lng),
    count: '1',
    language: 'fr',
    format: 'json',
  });

  const response = await fetch(`${BASE_URL}/search?${params}`);
  if (!response.ok) return null;

  const json = await response.json();
  const result: GeoLocation = json.results?.[0] ?? null;
  if (result) cacheSet(CACHE_KEYS.GEOCODING, cacheKey, result);
  return result;
}

export function locationKey(loc: GeoLocation): string {
  return `${loc.latitude.toFixed(4)}_${loc.longitude.toFixed(4)}`;
}

File: src-tauri/src/geocoding.rs
use serde_json::{json, Value};
use urlencoding;

pub async fn search_locations(query: &str) -> Result<Vec<Value>, String> {
    let url = format!(
        "https://geocoding-api.open-meteo.com/v1/search?name={}&count=10&language=fr",
        urlencoding::encode(query)
    );

    let client = reqwest::Client::new();
    let response = client
        .get(&url)
        .send()
        .await
        .map_err(|e| format!("Request failed: {}", e))?;

    let data: Value = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse response: {}", e))?;

    let results = data
        .get("results")
        .and_then(|r| r.as_array())
        .ok_or("No results found")?;

    let formatted: Vec<Value> = results
        .iter()
        .enumerate()
        .map(|(idx, loc)| {
            json!({
                "id": format!("{}", idx),
                "name": loc.get("name").and_then(|v| v.as_str()).unwrap_or(""),
                "admin1": loc.get("admin1").and_then(|v| v.as_str()),
                "country": loc.get("country").and_then(|v| v.as_str()).unwrap_or(""),
                "latitude": loc.get("latitude").and_then(|v| v.as_f64()),
                "longitude": loc.get("longitude").and_then(|v| v.as_f64()),
                "timezone": loc.get("timezone").and_then(|v| v.as_str()),
                "population": loc.get("population").and_then(|v| v.as_u64()),
            })
        })
        .collect();

    Ok(formatted)
}
